action.php

```diff
@@ -1,341 +1,348 @@
+				$result = array(
+					'message' => __( 'Could not find any autoresponders!', 'wp_smaily' ),
+					'error'   => true,
+				);
+				break;
+			}
 
-		break;
+			case 'refreshAutoresp':
+				global $wpdb;
+
+				// Load configuration data.
+				$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
+				$data       = $wpdb->get_row( "SELECT * FROM `$table_name` LIMIT 1" );
+
+				// Credentials.
+				$api_credentials = explode( ':', $data->api_credentials );
+				// Get autoresponders.
+				$request = new Smaily_Plugin_Request(
+					'https://' . $data->domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted',
+					array(
+						'username' => $api_credentials[0],
+						'password' => $api_credentials[1],
+					)
+				);
+				$result        = $request->get();
+				$autoresponders = $result['body'];
+
+				// Handle errors.
+				if ( isset( $result['code'] ) && $result['code'] !== 200 ) {
+					$result['error'] = true;
+					break;
+				} elseif ( empty( $autoresponders ) ) {
+					$result = array(
+						'message' => __( 'Could not find any autoresponders!', 'wp_smaily' ),
+						'error'   => true,
+					);
+					break;
+				}
+
+			$table_name = $wpdb->prefix . 'smaily_autoresponders';
+			$wpdb->query( "DELETE FROM `$table_name`" );
+			$wpdb->query( "INSERT INTO `$table_name`(`id`, `title`) VALUES " . implode( ',', $insert_query ) );
 
-	case 'save':
-		global $wpdb;
+			// Return result.
+			$result = array(
+				'error' => false,
+				'message' => __( 'Autoresponders refreshed.', 'wp_smaily' ),
+			);
 
-		// Get params.
-		$isAdvanced = ( isset( $_POST['is_advanced'] ) && ! empty( $_POST['is_advanced'] ) ) ? '1' : '0';
+			break;
 
-		// Get basic and advanced parameters.
-		$basic    = ( isset( $_POST['basic'] ) && is_array( $_POST['basic'] ) ) ? $_POST['basic'] : array();
-		$advanced = ( isset( $_POST['advanced'] ) && is_array( $_POST['advanced'] ) ) ? $_POST['advanced'] : array();
+		case 'save':
+			global $wpdb;
+			// Get parameters.
+			$isAdvanced = ( isset( $form_data['is_advanced'] ) && ! empty( $form_data['is_advanced'] ) ) ? '1' : '0';
 
-		// Generate new form (if empty).
-		if ( empty( $advanced['form'] ) ) {
-			require_once( BP . DS . 'code' . DS . 'Template.php' );
-			$template = new Wp_Sendsmaily_Template( 'html' . DS . 'form' . DS . 'advanced.php' );
+			// Get basic and advanced parameters.
+			$basic    = ( isset( $form_data['basic'] ) && is_array( $form_data['basic'] ) ) ? $form_data['basic'] : array();
+			$advanced = ( isset( $form_data['advanced'] ) && is_array( $form_data['advanced'] ) ) ? $form_data['advanced'] : array();
 
-			// Load configuration data.
-			$table_name = esc_sql( $wpdb->prefix . 'sendsmaily_config' );
-			$data = $wpdb->get_row( "SELECT * FROM `$table_name` LIMIT 1" );
-			$template->assign( (array) $data );
+			// Validate and sanitize basic & advanced parameters values.
+			$autoresponder = ( isset( $basic['autoresponder'] ) && is_int( $basic['autoresponder'] ) )
```

Comment:
that is correct. what would be the best way to validate only whole numbers in this case? should i just convert $basic['autoresponder'] to int or use is_numeric?

Reply:
There is an issue with your updated implementation - `(int) 'abc' === 0`. Can you explain why this is a bad thing?

Reply:
I thought doing this was bad because strings would cast to 0.
But if autoresponder will be saved as 0 then the admin page will display "no autoresponder set"

Reply:
It is fine to cast string into 0, but the issue is that you are checking if `(int) $basic['autoresponder']` is an integer (which it always will be), but you are assigning the raw value to `$autoresponder` variable.

Basically `; DROP DATABASE wordpress` would still be allowed through normalization. Granted, `wpdb::prepare` will sanitize the value, but it is not guaranteed to be safe.

Reply:
Checking if is_int is pointless when I've already casted, I'll fix it. Doesnt casting sanitize DROP DATABASE into 0 before prepare is execured? Or is it just bad practice to sanitize this way?

Reply:
Yes, casting `; DROP DATABASE wordpress` will result in `0`.

It depends on the context if casting value to integer is a bad practice or not. For example if variable value `0` has a meaning in some sort of business logic, then it is a bad practice, because invalid value (for example `foo`) will be silently converted to `0` and that can lead to plethora of issues down the line.

The rule is - be explicit about what type of values you are expecting. If variable value can be either a autoresponder ID (integer) or no selection (empty string), then only accept those values, reject everything else with an error.